Buttplug.Server.Managers.ErosTek/ET312Device.cs

```diff
@@ -0,0 +1,398 @@
+
+                // If the response is not of the expected type or Checksum doesn't match
+                // consider ourselves de-synchronized. Calling Code should treat the device
+                // as disconnected
+                if (recBuffer[0] != 0x22 || recBuffer[2] != Checksum(recBuffer))
+                {
+                    throw new ET312CommunicationException();
+                }
+
+                return recBuffer[1];
+            }
+        }
+
+        // TODO: Replace this by some kind of asynchronous processing so waiting for
+        //       the response from the box doesn't hold up the entire event loop
+        private void Poke(uint address, byte value)
+        {
+            lock (serialPort)
+            {
+                byte[] sendBuffer = new byte[5];
+                sendBuffer[0] = 0x4d;                               // write byte command
+                sendBuffer[1] = (byte)((address & 0xff00) >> 8);    // address high byte
+                sendBuffer[2] = (byte)(address & 0x00ff);           // address low byte
+                sendBuffer[3] = value;                              // value
+                sendBuffer[4] = Checksum(sendBuffer);               // checksum
+                SillyXOR(sendBuffer);                               // encryption
+
+                serialPort.Write(sendBuffer, 0, 5);
+
+                // If the response is not ACK, consider ourselves de-synchronized.
+                // Calling Code should treat the device as disconnected
+                if (serialPort.ReadByte() != 0x06)
+                {
+                    throw new ET312CommunicationException();
+                }
+            }
+        }
+
+        // Execute box command
+        private void Execute(byte command)
+        {
+            Poke(0x4070, command);
+            System.Threading.Thread.Sleep(18);
+        }
+
+        // Write Message to the LCD display
+        private void WriteLCD(string message, int offset)
+        {
+            byte[] buffer = Encoding.ASCII.GetBytes(message);
+            for (int c = 0; c < buffer.Length; c++)
+            {
+                Poke(0x4180, buffer[c]);              // ASCII Value
+                Poke(0x4181, (byte)(offset + c));     // Offset (+64 = second line)
+                Execute(0x13);                        // Write Character
+            }
+        }
+
+        // Set up serial XOR encryption - mandatory for write accesss
+        private void Synch()
+        {
+            lock (serialPort)
+            {
+                BpLogger.Info("Attempting Synch");
+
+                try
+                {
+                    // Try to read a byte from RAM using an unencrypted command
+                    // If this works, we are not synched yet
+                    byte[] sendBuffer = new byte[4];
+                    sendBuffer[0] = 0x3c;                               // read byte command
```

Comment:
Magic numbers require manual look up to confirm what's going on. use `enum`s instead